VMS: Added new method to gather entropy on VMS, based on SYS$GET_ENTROPY

levitte opened the pull request:
This system services is based on FreeBSD 12's getentropy(), and is
therefore treated the same way as getentropy() with regards to amount
of entropy bits per data bit.

levitte committed:
VMS: Added new method to gather entropy on VMS, based on SYS$GET_ENTROPY.

This system services is based on FreeBSD 12's getentropy(), and is
therefore treated the same way as getentropy() with regards to amount
of entropy bits per data bit.

levitte commented:
Note: I'm keeping this a draft until VSI has started rolling out the patch that adds this system service.

t-j-h commented:
Putting WIP in the title helps as draft doesn't show in the email thread only on the web interface.

mspncp commented:
`s/gether/gather/g`  (commit message and pr title)

mspncp commented on CHANGES:
```diff
@@ -9,6 +9,10 @@
 
  Changes between 1.1.1 and 3.0.0 [xx XXX xxxx]
 
+  *) Added new method to gether entropy on VMS, based on SYS$GET_ENTROPY.
```
And here, too.

levitte replied:
Fixed

paulidale commented on CHANGES:
```diff
@@ -9,6 +9,10 @@
  Changes between 1.1.1 and 3.0.0 [xx XXX xxxx]
 
+  *) Added new method to gether entropy on VMS, based on SYS$GET_ENTROPY.
+     The presence of that system service is done in run-time.
```
`s/that/this/`
`s/done in/performed at/`

levitte replied:
... and `s/gether/gather/`...

paulidale commented on crypto/rand/rand_vms.c:
```diff
@@ -32,10 +32,21 @@
+ * DATA COLLECTION METHOD
+ * ======================
+ *
+ * This is a very low entropy method to get *some* entropy.
```
This is a method to get *some* very low quality entropy ??

levitte replied:
Much better phrasing, thank you.  Applied

paulidale commented on crypto/rand/rand_vms.c:
```diff
@@ -497,6 +508,73 @@ int rand_pool_add_nonce_data(RAND_POOL *pool)
+ *
+ * This is a high entropy method based on a new systems service that is
+ * based on getentropy() from FreeBSD 12.  It's only used if available,
+ * and its availability is detected in run-time.
```
s/in/at/

levitte replied:
The mysteries of English.  Fixed

paulidale commented on crypto/rand/rand_vms.c:
```diff
@@ -497,6 +508,73 @@ int rand_pool_add_nonce_data(RAND_POOL *pool)
+{
+    if (get_entropy_address_flag == 0) {
+        get_entropy_addr = dlsym(dlopen(PUBLIC_VECTORS, 0), OLD_ROUTINE);
+    }
```
Lose the {}'s ??

levitte replied:
Okie

paulidale commented on crypto/rand/rand_vms.c:
```diff
@@ -497,6 +508,73 @@ int rand_pool_add_nonce_data(RAND_POOL *pool)
+size_t get_entropy_method(RAND_POOL *pool)
+{
+    /*
+     * Documentation says that SYS$GET_ENTROPY will give a maximum of
```
`The documentation ...`

paulidale commented on crypto/rand/rand_vms.c:
```diff
@@ -497,6 +508,73 @@ int rand_pool_add_nonce_data(RAND_POOL *pool)
+     * 256 bytes of data.
+     */
+    unsigned char buffer[256];
+    size_t bytes_needed = rand_pool_bytes_needed(pool, 1);
```
Is a range check against 256 appropriate?
I realise that 2048 bits seems unlikely....

levitte replied:
I use `bytes_remaining` in the call to the system service, and if that's too big, it will return a status that says so (i.e. Something other than `SS$_NORMAL` or `SS$_RETRY`.

paulidale reviewed (commented):
A few nits for the most part.

levitte committed:
fixup! VMS: Added new method to gather entropy on VMS, based on SYS$GET_ENTROPY.

levitte committed:
fixup! VMS: Added new method to gather entropy on VMS, based on SYS$GET_ENTROPY.

mspncp commented on crypto/rand/rand_vms.c:
```diff
@@ -497,6 +502,72 @@ int rand_pool_add_nonce_data(RAND_POOL *pool)
+ * SYS$GET_ENTROPY METHOD
+ * ======================
+ *
+ * This is a high entropy method based on a new systems service that is
```
system~s~ service

levitte replied:
Okie

mspncp commented on crypto/rand/rand_vms.c:
```diff
@@ -497,6 +502,72 @@ int rand_pool_add_nonce_data(RAND_POOL *pool)
+ * ======================
+ *
+ * This is a high entropy method based on a new systems service that is
+ * based on getentropy() from FreeBSD 12.  It's only used if available,
```
based on the getentropy() system call

levitte replied:
Whether `getentropy()` is a system call or a library function seems to vary, even between BSD branches, so let's avoid that qualifier, yeah?

mspncp replied:
okidoki

mspncp commented on crypto/rand/rand_vms.c:
```diff
@@ -497,6 +502,72 @@ int rand_pool_add_nonce_data(RAND_POOL *pool)
+ * based on getentropy() from FreeBSD 12.  It's only used if available,
+ * and its availability is detected at run-time.
+ *
+ * We assume that this function gives 1 bit of entropy for each bit of
```
We assume that this function provides full entropy random output.

levitte replied:
Sure

mspncp commented on crypto/rand/rand_vms.c:
```diff
@@ -32,10 +32,21 @@
+ *
+ * This is a method to get *some* very low quality entropy.
+ * It works by collecting all kinds of statistical data that
+ * VMS offers and using them as "entropy".
```
Placing terms in double quotes gives them a dubious emphasis (at that's the case in german language.)

```
... and using them as random seed.
```

levitte replied:
Sure

mspncp reviewed (commented):
Not knowing much about VMS, I'm only able to pick on some typos and suggest some alternative formulations. Many of the suggestions are a matter of personal taste, feel free to adopt what you like and  reject what you don't.

levitte committed:
fixup! VMS: Added new method to gather entropy on VMS, based on SYS$GET_ENTROPY.

levitte commented:
I've tested this on VMS with the 1.1.1 branch (the master branch is currently in a pitiful state on VMS)

levitte commented on CHANGES:
```diff
@@ -9,6 +9,10 @@
 
  Changes between 1.1.1 and 3.0.0 [xx XXX xxxx]
 
+  *) Added new method to gather entropy on VMS, based on SYS$GET_ENTROPY.
```
Okie

levitte commented on crypto/rand/rand_vms.c:
```diff
@@ -497,6 +502,72 @@ int rand_pool_add_nonce_data(RAND_POOL *pool)
+}
+
+/*
+ * MAIN ENTTROPY AQUISITION FUNCTIONS
```
Thanks

levitte commented on crypto/rand/rand_vms.c:
```diff
@@ -32,10 +32,21 @@
+ * DATA COLLECTION METHOD
+ * ======================
+ *
+ * This is a method to get *some* very low quality entropy.
```
I actually did some measurements, among others using the NIST test suite...  so while still not entirely objective, it's more of an observation than a mere gut feeling.

levitte replied:
... but ok

levitte committed:
fixup! VMS: Added new method to gather entropy on VMS, based on SYS$GET_ENTROPY.

levitte commented:
Ping!

paulidale commented on crypto/rand/rand_vms.c:
```diff
@@ -497,6 +502,80 @@ int rand_pool_add_nonce_data(RAND_POOL *pool)
+    for (bytes_needed = rand_pool_bytes_needed(pool, 1);
+         bytes_needed > 0;
+         bytes_needed -= bytes_to_get) {
+        bytes_to_get =
```
Indent?

paulidale reviewed (approved):
Looks plausible to me.

One indent nit.

paulidale commented:
Should this go back to 1.1.1 as well?
Better entropy is a bug fix in my opinion.

levitte commented:
> Should this go back to 1.1.1 as well?
> Better entropy is a bug fix in my opinion.

I've argued the same thing, but others disagree.  I can go either way, so someone else decide, please.

levitte commented:
It seems we have enough quorum for a backport, see the label that got added just now

paulidale reviewed (approved):
Good for 1.1.1 too.

levitte commented:
Merged.

master:
8b9896eb293a0861f0b8c191b7a278f176b729e6 VMS: Added new method to gather entropy on VMS, based on SYS$GET_ENTROPY.

1.1.1:
0a71b62107e7175d9bbecf9458bbc0bcbf05d148 VMS: Added new method to gather entropy on VMS, based on SYS$GET_ENTROPY.
